lockbox-ios/Store/AutoLockStore.swift

```diff
@@ -0,0 +1,114 @@
+            } else {
+                self.setupTimer()
+            }
+        }).disposed(by: self.disposeBag)
+
+        self.dispatcher.register
+            .filterByType(class: LifecycleAction.self)
+            .subscribe(self.lifecycleAction)
+            .disposed(by: self.disposeBag)
+    }
+
+    private func lifecycleAction(evt: Event<LifecycleAction>) {
+        if evt.element == .background {
+            self.userDefaults.onAutoLockTime
+                .take(1)
+                .subscribe(onNext: { (latest: AutoLockSetting) in
+                    if latest == AutoLockSetting.OnAppExit {
+                        self.lockApp()
+                    }
+                }).disposed(by: self.disposeBag)
+        }
+    }
+
+    private func setupTimer() {
+        self.userDefaults.onAutoLockTime
+            .take(1)
+            .subscribe(onNext: { (latest: AutoLockSetting) in
+                switch latest {
+                case .FiveMinutes:
+                    self.setTimer(seconds: 60 * 5)
+                case .OneHour:
+                    self.setTimer(seconds: 60 * 60)
+                case .OneMinute:
+                    self.setTimer(seconds: 60)
+                case .TwelveHours:
+                    self.setTimer(seconds: 60 * 60 * 12)
+                case .TwentyFourHours:
+                    self.setTimer(seconds: 60 * 60 * 24)
+                case .Never:
+                    self.stopTimer()
+                case .OnAppExit:
+                    return
+                }
+            })
+            .disposed(by: self.disposeBag)
+    }
+
+    private func setTimer(seconds: Int) {
+        let timerValue = self.userDefaults.double(forKey: SettingKey.autoLockTimerDate.rawValue)
+        if timerValue != 0 {
+            self.timer = Timer(fireAt: Date(timeIntervalSince1970: timerValue),
+                               interval: 0,
+                               target: self,
+                               selector: #selector(lockApp),
+                               userInfo: nil,
+                               repeats: false)
+        } else {
+            self.timer = Timer(timeInterval: TimeInterval(seconds),
+                               target: self,
+                               selector: #selector(lockApp),
+                               userInfo: nil,
+                               repeats: false)
+
+            self.userDefaults.set(self.timer?.fireDate.timeIntervalSince1970,
+                                  forKey: SettingKey.autoLockTimerDate.rawValue)
+        }
+    }
+
+    private func stopTimer() {
+        if let timer = self.timer {
+            timer.invalidate()
+        }
+
+        self.userDefaults.removeObject(forKey: SettingKey.autoLockTimerDate.rawValue)
+    }
+
+    @objc private func lockApp() {
+        self.settingActionHandler.invoke(SettingAction.visualLock(locked: true))
+        self.routeActionHandler.invoke(LoginRouteAction.welcome)
```

Comment:
I think the `routeActionHandler` call is extraneous, we do that routing in the `RootPresenter` based on the `visualLock` setting anyway.